File: code/python/rest/client/task_api_client.py
import asyncio
from typing import Dict, List, Optional, Any
from datetime import datetime
import httpx
import json


class TaskAPIClient:
    """REST API client for Task Management Service"""
    
    def __init__(self, base_url: str = "http://localhost:8000/api/v1"):
        self.base_url = base_url
        self.headers = {
            "Content-Type": "application/json",
            "Accept": "application/json"
        }
        self.client = httpx.AsyncClient(
            base_url=base_url,
            headers=self.headers,
            timeout=30.0
        )
    
    async def __aenter__(self):
        return self
    
    async def __aexit__(self, exc_type, exc_val, exc_tb):
        await self.close()
    
    async def close(self):
        """Close the HTTP client"""
        await self.client.aclose()
    
    def set_auth_token(self, token: str):
        """Set authorization token"""
        self.headers["Authorization"] = f"Bearer {token}"
        self.client.headers.update(self.headers)
    
    def set_api_key(self, api_key: str):
        """Set API key"""
        self.headers["X-API-Key"] = api_key
        self.client.headers.update(self.headers)
    
    async def list_tasks(
        self,
        page_size: int = 20,
        page_token: Optional[str] = None,
        status: Optional[str] = None,
        assigned_to: Optional[str] = None,
        tags: Optional[str] = None,
        sort_order: Optional[str] = None
    ) -> Dict[str, Any]:
        """List tasks with optional filters"""
        params = {
            "page_size": page_size
        }
        
        if page_token:
            params["page_token"] = page_token
        if status:
            params["status"] = status
        if assigned_to:
            params["assigned_to"] = assigned_to
        if tags:
            params["tags"] = tags
        if sort_order:
            params["sort_order"] = sort_order
        
        response = await self.client.get("/tasks", params=params)
        response.raise_for_status()
        return response.json()
    
    async def get_task(self, task_id: str) -> Dict[str, Any]:
        """Get a single task by ID"""
        response = await self.client.get(f"/tasks/{task_id}")
        
        if response.status_code == 404:
            raise TaskNotFoundError(f"Task with ID {task_id} not found")
        
        response.raise_for_status()
        return response.json()
    
    async def create_task(
        self,
        title: str,
        description: Optional[str] = None,
        priority: Optional[str] = None,
        tags: Optional[List[str]] = None,
        assigned_to: Optional[str] = None,
        due_date: Optional[datetime] = None
    ) -> Dict[str, Any]:
        """Create a new task"""
        data = {
            "title": title
        }
        
        if description is not None:
            data["description"] = description
        if priority:
            data["priority"] = priority
        if tags:
            data["tags"] = tags
        if assigned_to:
            data["assigned_to"] = assigned_to
        if due_date:
            data["due_date"] = due_date.isoformat() if isinstance(due_date, datetime) else due_date
        
        response = await self.client.post("/tasks", json=data)
        
        if response.status_code == 400:
            error = response.json()
            raise ValidationError(error.get("error", {}).get("message", "Validation failed"))
        
        response.raise_for_status()
        return response.json()
    
    async def update_task(
        self,
        task_id: str,
        title: Optional[str] = None,
        description: Optional[str] = None,
        status: Optional[str] = None,
        priority: Optional[str] = None,
        tags: Optional[List[str]] = None,
        assigned_to: Optional[str] = None,
        due_date: Optional[datetime] = None
    ) -> Dict[str, Any]:
        """Update an existing task"""
        data = {}
        
        if title is not None:
            data["title"] = title
        if description is not None:
            data["description"] = description
        if status is not None:
            data["status"] = status
        if priority is not None:
            data["priority"] = priority
        if tags is not None:
            data["tags"] = tags
        if assigned_to is not None:
            data["assigned_to"] = assigned_to
        if due_date is not None:
            data["due_date"] = due_date.isoformat() if isinstance(due_date, datetime) else due_date
        
        response = await self.client.put(f"/tasks/{task_id}", json=data)
        
        if response.status_code == 404:
            raise TaskNotFoundError(f"Task with ID {task_id} not found")
        if response.status_code == 400:
            error = response.json()
            raise ValidationError(error.get("error", {}).get("message", "Validation failed"))
        
        response.raise_for_status()
        return response.json()
    
    async def update_task_status(self, task_id: str, status: str) -> Dict[str, Any]:
        """Update task status"""
        data = {"status": status}
        
        response = await self.client.patch(f"/tasks/{task_id}/status", json=data)
        
        if response.status_code == 404:
            raise TaskNotFoundError(f"Task with ID {task_id} not found")
        if response.status_code == 400:
            error = response.json()
            raise ValidationError(error.get("error", {}).get("message", "Validation failed"))
        
        response.raise_for_status()
        return response.json()
    
    async def delete_task(self, task_id: str) -> bool:
        """Delete a task"""
        response = await self.client.delete(f"/tasks/{task_id}")
        
        if response.status_code == 404:
            raise TaskNotFoundError(f"Task with ID {task_id} not found")
        
        response.raise_for_status()
        return response.status_code == 204


class TaskAPIError(Exception):
    """Base exception for Task API errors"""
    pass


class TaskNotFoundError(TaskAPIError):
    """Task not found error"""
    pass


class ValidationError(TaskAPIError):
    """Validation error"""
    pass


def print_task_list(tasks: List[Dict[str, Any]]):
    """Helper function to print formatted task list"""
    for task in tasks:
        status = task['status'].upper()
        priority = task['priority'].upper()
        title = task['title']
        
        print(f"[{status}] {title} (Priority: {priority})")
        
        if task.get('description'):
            print(f"  Description: {task['description']}")
        
        if task.get('tags'):
            print(f"  Tags: {', '.join(task['tags'])}")
        
        if task.get('assigned_to'):
            print(f"  Assigned to: {task['assigned_to']}")
        
        if task.get('due_date'):
            print(f"  Due: {task['due_date']}")
        
        print()


async def main():
    """Example usage of the Task API client"""
    async with TaskAPIClient() as client:
        try:
            # Create a task
            print("Creating task...")
            task = await client.create_task(
                title="Test Python Client",
                description="Testing the Python REST client",
                priority="high",
                tags=["test", "python"]
            )
            task_id = task['id']
            print(f"Created task: {task_id}")
            
            # List tasks
            print("\nListing tasks...")
            result = await client.list_tasks(page_size=10, sort_order="priority_desc")
            print_task_list(result['tasks'])
            
            # Update task status
            print("Updating task status...")
            updated = await client.update_task_status(task_id, "in_progress")
            print(f"Task status updated to: {updated['status']}")
            
            # Get single task
            print("\nFetching task details...")
            fetched = await client.get_task(task_id)
            print(f"Task: {fetched['title']} - Status: {fetched['status']}")
            
            # Delete task
            print("\nDeleting task...")
            if await client.delete_task(task_id):
                print("Task deleted successfully")
            
        except TaskAPIError as e:
            print(f"Error: {e}")
        except Exception as e:
            print(f"Unexpected error: {e}")


if __name__ == "__main__":
    asyncio.run(main())

File: code/ruby/rest/client/task_api_client.rb
require 'httparty'
require 'json'

class TaskAPIClient
  include HTTParty
  
  def initialize(base_url = 'http://localhost:4567/api/v1')
    @base_url = base_url
    self.class.base_uri @base_url
    @headers = {
      'Content-Type' => 'application/json',
      'Accept' => 'application/json'
    }
  end
  
  def set_auth_token(token)
    @headers['Authorization'] = "Bearer #{token}"
  end
  
  def set_api_key(api_key)
    @headers['X-API-Key'] = api_key
  end
  
  def list_tasks(page_size: nil, page_token: nil, status: nil, assigned_to: nil, tags: nil, sort_order: nil)
    query = {}
    query[:page_size] = page_size if page_size
    query[:page_token] = page_token if page_token
    query[:status] = status if status
    query[:assigned_to] = assigned_to if assigned_to
    query[:tags] = tags if tags
    query[:sort_order] = sort_order if sort_order
    
    response = self.class.get('/tasks', 
      headers: @headers,
      query: query
    )
    
    handle_response(response)
  end
  
  def get_task(id)
    response = self.class.get("/tasks/#{id}", headers: @headers)
    handle_response(response)
  end
  
  def create_task(title:, description: nil, priority: nil, tags: nil, assigned_to: nil, due_date: nil)
    body = {
      title: title,
      description: description,
      priority: priority,
      tags: tags,
      assigned_to: assigned_to,
      due_date: due_date
    }.compact
    
    response = self.class.post('/tasks',
      headers: @headers,
      body: body.to_json
    )
    
    handle_response(response)
  end
  
  def update_task(id, **updates)
    response = self.class.put("/tasks/#{id}",
      headers: @headers,
      body: updates.to_json
    )
    
    handle_response(response)
  end
  
  def update_task_status(id, status)
    response = self.class.patch("/tasks/#{id}/status",
      headers: @headers,
      body: { status: status }.to_json
    )
    
    handle_response(response)
  end
  
  def delete_task(id)
    response = self.class.delete("/tasks/#{id}", headers: @headers)
    
    if response.code == 204
      true
    else
      handle_response(response)
    end
  end
  
  private
  
  def handle_response(response)
    case response.code
    when 200, 201
      response.parsed_response
    when 204
      nil
    when 400
      raise ValidationError, "Validation failed: #{response.parsed_response['error']['message']}"
    when 404
      raise NotFoundError, "Resource not found: #{response.parsed_response['error']['message']}"
    when 401
      raise AuthenticationError, "Authentication failed"
    when 403
      raise AuthorizationError, "Access denied"
    when 500
      raise ServerError, "Server error: #{response.parsed_response['error']['message']}"
    else
      raise APIError, "Unexpected response: #{response.code}"
    end
  end
  
  class APIError < StandardError; end
  class ValidationError < APIError; end
  class NotFoundError < APIError; end
  class AuthenticationError < APIError; end
  class AuthorizationError < APIError; end
  class ServerError < APIError; end
end

# Example usage
if __FILE__ == $0
  client = TaskAPIClient.new
  
  begin
    # Create a task
    puts "Creating task..."
    task = client.create_task(
      title: 'Test Ruby Client',
      description: 'Testing the Ruby REST client',
      priority: 'high',
      tags: ['test', 'ruby']
    )
    puts "Created task: #{task['id']}"
    
    # List tasks
    puts "\nListing tasks..."
    result = client.list_tasks(page_size: 10)
    result['tasks'].each do |t|
      puts "  [#{t['status'].upcase}] #{t['title']} (Priority: #{t['priority']})"
    end
    
    # Update task status
    puts "\nUpdating task status..."
    updated = client.update_task_status(task['id'], 'in_progress')
    puts "Task status updated to: #{updated['status']}"
    
    # Get single task
    puts "\nFetching task details..."
    fetched = client.get_task(task['id'])
    puts "Task: #{fetched['title']} - Status: #{fetched['status']}"
    
    # Delete task
    puts "\nDeleting task..."
    client.delete_task(task['id'])
    puts "Task deleted successfully"
    
  rescue => e
    puts "Error: #{e.message}"
  end
end